docs/deprecation_policy.md

```diff
@@ -0,0 +1,167 @@
+version v1 (deprecated) but not in v2alpha1.  The Widget resource continues to
+exist and function in releases up to and including X+8.  Only in release X+9,
+when API v1 has aged out, does the Widget resource cease to exist, and the
+behavior get removed.
+
+### Fields of REST resources
+
+As with whole REST resources, an individual field which was present in API v1
+must exist and function until API v1 is removed.  Unlike whole resources, the
+v2 APIs may choose a different representation for the field, as long as it can
+be round-tripped.  For example a v1 field named “magnitude” which was
+deprecated might be named “deprecatedMagnitude” in API v2.  When v1 is
+eventually removed, the deprecated field can be removed from v2.
+
+### Enumerated or constant values
+
+As with whole REST resources and fields thereof, a constant value which was
+supported in API v1 must exist and function until API v1 is removed.
+
+### Future work
+
+We intend to introduce more fine-grained API versions, at which point we will
+adjust these rules as needed.
+
+## Deprecating a flag or CLI
+
+The Kubernetes system is comprised of several different programs cooperating.
+Sometimes, we need to remove flags or CLI commands (collectively “CLI
+elements”) in these programs.  The individual programs naturally sort into two
+main groups - user-facing and admin-facing programs, which vary slightly in
+their deprecation policies.  Unless a flag is explicitly prefixed “alpha” or
+“beta”, it is considered GA.
+
+CLI elements are effectively part of the API to the system, but since they are
+not versioned in the same way as the REST API, the rules for deprecation are as
+follows:
+
+Rule #4a: CLI elements of user-facing components (e.g. kubectl) must function
+after their announced deprecation for no less than:
+   * GA: 1 year or 2 releases (whichever is longer)
+   * Beta: 3 months or 1 release (whichever is longer)
+   * Alpha: 0 releases
+
+Rule #4b: CLI elements of admin-facing components (e.g. kubelet) must function
+after their announced deprecation for no less than:
+   * GA: 6 months or 1 release (whichever is longer)
+   * Beta: 3 months or 1 release (whichever is longer)
+   * Alpha: 0 releases
+
+Rule #5: Deprecated CLI elements must emit warnings (optionally disableable)
+when used.
+
+## Deprecating a feature or behavior
+
+Occasionally we need to deprecate some feature or behavior of the system that
+is not controlled by the API or CLI.  In this case, the rules for deprecation
+are as follows:
+
+Rule #6: Deprecated behaviors must function for no less than 1 year after their
+announced deprecation.
+
+This does not imply that all changes to the system are governed by this policy.  This applies only to significant, user-visible behaviors which impact the correctness of applications running on Kubernetes or that impact the administration of Kubernetes clusters, and which are being removed entirely.
```

Comment:
This subjective policy will lead to confusion. Would it make sense to define a process for deciding deprecation period for such features?